### PR TITLE
[RTC-87] Elixir server SDK

### PR DESCRIPTION
**Pull request by LVala**
At the moment test are made using mocks, but in the future we can test the SDK with actual Jellyfish server. Also, it's worth noting that implementations for `Component` and `Peer` are very similar, but I figured the features may diverge in some time.

### CODE REVIEW

**Review by Rados13 — APPROVED** [review]

**Review by mickel8 — COMMENTED** [review]
I just took a look and would propose following hierarchy


```elixir
# SDK level
Jellyfish.SDK.Client.new
Jellyfish.SDK.Client.create_room
Jellyfish.SDK.Client.delete_room
Jellyfish.SDK.Client.get_rooms
Jellyfish.SDK.Client.get_room_by_id

# Room level
Jellyfish.SDK.Room.add_peer
Jellyfish.SDK.Room.delete_peer
Jellyfish.SDK.Room.add_component
Jellyfish.SDK.Room.delete_component
Jellyfish.SDK.Room.get_componets
Jellyfish.SDK.Room.get_peers
```

or move everything on top level

**Comment by mickel8** [comment]
Assigned also @bblaszkow06 and @daniel-jodlos to do more precies CR (if they have time) and make familiar with the concept

File: lib/jellyfish/sdk/room.ex [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  @doc ~S"""
+  Sends request to delete specified room.
+
+  ## Parameters
+
+    * `client` - instance of `t:Jellyfish.SDK.Client.t/0`
+    * `room_id` - id of the room that will be deleted
+  """
+  @spec delete_room(Client.t(), String.t()) :: :ok | {:error, String.t()}
+  def delete_room(client, room_id) do
```
Comment (daniel-jodlos) [inline lib/jellyfish/sdk/room.ex:92]:
```suggestion
  def delete(client, room_id) do
```

File: lib/jellyfish/sdk/room.ex [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+  @doc ~S"""
+  Sends request for specified room metadata.
+
+  ## Parameters
+
+    * `client` - instance of `t:Jellyfish.SDK.Client.t/0`
+  """
+  @spec get_rooms(Client.t()) :: {:ok, [t()]} | {:error, String.t()}
+  def get_rooms(client) do
```
Comment (daniel-jodlos) [inline lib/jellyfish/sdk/room.ex:107]:
```suggestion
  def list(client) do
```
or
```suggestion
  def list_all(client) do
```

File: lib/jellyfish/sdk/room.ex (hunk 1 above)
Comment (daniel-jodlos) [reply lib/jellyfish/sdk/room.ex:92]: I don't have a definitive answer for this question, but I'm wondering if these functions should take 2 arguments or a single Room struct that would also contain a client ref. These two will always be passed around together on the BE, especially if the BE manages multiple instances of JF, and majority of actions will take place on the room level. Even if they're not, you cannot do anything without the client ref.

I was basically thinking about something that would end up looking more less like this:
```elixir
{:ok, client} = Client.connect("localhost")
{:ok, room} = Client.Room.create(client, "room", max_peers: 10)
{:ok, component} = Client.Room.add_component(room, :hls, id: "hls")
:ok = Client.Room.Component.remove(component)
```

**Review by daniel-jodlos — CHANGES_REQUESTED** [review]

File: lib/jellyfish/sdk/component.ex [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  Maps a `"data"` field of request response body from string keys to atom keys.
+
+  Will fail if the input structure is invalid.
+
+  ## Parameters
+
+    * `response` - a map representing JSON response
+  """
+  @spec component_from_json(map()) :: t()
+  def component_from_json(response) do
```
[the remark this answers is not in the document]
Comment (LVala) [reply lib/jellyfish/sdk/component.ex:100]: I made these functions private, also it didn't return a tuple because I used it in `Enum.map`. I don't think that it's a problem now when it's private.

[the remark this answers is not in the document]
Comment (daniel-jodlos) [reply lib/jellyfish/sdk/component.ex]: It has nothing to do with the function being private and everything to do with the fact that it can throw a `BadMatch`

**Comment by LVala** [comment]
I mostly changed under what module functions are and documentation (I'm still not sure about some aspects so please take a look at it). Also, didn't fix the tests yet as there might be some more changes.

File: lib/jellyfish/sdk/component.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (LVala) [reply lib/jellyfish/sdk/component.ex]: By it being private I meant that now only we have to care about using it correctly. I can rewrite it to return `{:ok | :error, _}` tuple, but it makes it very painful to use in `Enum.map`, I already discussed it with @mickel8 when I was writing it.

[the remark this answers is not in the document]
Comment (LVala) [reply lib/jellyfish/sdk/component.ex]: It throwing error means that you use wrong combo of SDK/Jellyfish versions or somehow server replies have invalid structure. I agree that we should handle it better that throwing `BadMatch`, but I don't think that's an error that user should be able to recover from (well, now he cannot use the function anyway), but I may not understand something.

File: lib/jellyfish/component.ex [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  """
+
+  @enforce_keys [
+    :id,
+    :type
+  ]
+  defstruct @enforce_keys
+
+  @typedoc """
+  Id of a component in a form of UUID string.
```
Comment (daniel-jodlos) [inline lib/jellyfish/component.ex:13]: A bit too specific I think. It's also a little... revealing. This is the SDK mind you, it just makes API calls and it shouldn't care about server implementation details for our own sanity. We're going to have a whole bunch of these, and correcting server implementation details in docs for 20 SDKs doesn't sound like fun. But I think we should mention that the peer id is unique in the <room/JF (?)> scope.
```suggestion
  Id of the component, unique across every room
```

**Review by daniel-jodlos — CHANGES_REQUESTED** [review]

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,30 +1,25 @@
...
-          context:
-            - Deployment
-          filters:
-            branches:
-              ignore: /.*/
-            tags:
-              only: /v.*/
+      - elixir/build_test
+      - test
+      - elixir/lint
```
Comment (bblaszkow06) [inline .circleci/config.yml:25]: Was that intended to remove publish on the tag job?

File: mix.exs [hunk 6]
```diff
@@ -37,10 +48,15 @@ defmodule Membrane.Template.Mixfile do
...
+      {:tesla, "~> 1.5"},
+      {:hackney, "~> 1.18"},
+      {:jason, "~> 1.4"},
+
+      # Docs, credo, test coverage, dialyzer
       {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
       {:dialyxir, ">= 0.0.0", only: :dev, runtime: false},
-      {:credo, ">= 0.0.0", only: :dev, runtime: false}
+      {:credo, ">= 0.0.0", only: :dev, runtime: false},
+      {:excoveralls, "~> 0.15.0", only: :test, runtime: false}
```
Comment (bblaszkow06) [inline mix.exs:59]: Is there a reason to lock patch version of excoveralls? IMO it could be defined same way as `ex_doc`

File: test/jellyfish/room_test.exs [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+          method: :post,
+          url: "#{@url}/room",
+          body: "{\"maxPeers\":#{@max_peers_string}}"
+        } ->
+          json(%{"data" => build_room_json(true)}, status: 201)
+
+        %{
+          method: :post,
+          url: "#{@url}/room",
+          body: "{\"maxPeers\":\"#{@invalid_max_peers}\"}"
```
Comment (bblaszkow06) [inline test/jellyfish/room_test.exs:57]: I'd suggest using sigil `~s` with different delimiters, e.g. `|` to avoid escaping of double quotes. Example:

```suggestion
          body: ~s|{"maxPeers":"#{@invalid_max_peers}"}|
```

File: test/jellyfish/room_test.exs [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+        } ->
+          json(%{"errors" => @error_message}, status: 422)
```
Comment (bblaszkow06) [inline test/jellyfish/room_test.exs:59]: Passing string as `maxPeers` is rather something that qualifies for `400 Bad request` status code. See https://stackoverflow.com/a/52363900/10186815

File: test/jellyfish/room_test.exs [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+    end
+  end
+
+  describe "Room.add_component/4" do
+    setup do
+      mock(fn
+        %{
+          method: :post,
+          url: "#{@url}/room/#{@room_id}/component",
+          body: "{\"options\":{},\"type\":\"#{@component_type}\"}"
```
Comment (bblaszkow06) [inline test/jellyfish/room_test.exs:170]:
```suggestion
          body: ~s|{"options":{},"type":"#{@component_type}"}|
```

File: lib/jellyfish/utils.ex [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+  alias Tesla.Env
+
+  @spec translate_error_response({:error, any()}) :: {:error, String.t()}
+  def translate_error_response({:ok, %Env{body: %{"errors" => error}}}) do
+    {:error, "Request failed: #{error}"}
+  end
+
+  def translate_error_response({:ok, %Env{body: body}}) do
+    {:error, "Received unexpected response: #{inspect({body})}"}
```
Comment (bblaszkow06) [inline lib/jellyfish/utils.ex:12]: It doesn't seem like an error the client should handle
1. This should be distinguishable from other errors, so additional atom could be used
2. It's an internal SDK error and IMO this should raise

File: lib/jellyfish/utils.ex [hunk 11]
```diff
@@ -0,0 +1,18 @@
+defmodule Jellyfish.Utils do
+  @moduledoc false
+
+  alias Tesla.Env
+
+  @spec translate_error_response({:error, any()}) :: {:error, String.t()}
+  def translate_error_response({:ok, %Env{body: %{"errors" => error}}}) do
+    {:error, "Request failed: #{error}"}
```
Comment (bblaszkow06) [inline lib/jellyfish/utils.ex:8]: I'm not sure how "temporary" do you consider this API version, but IMO if you really want the error to "handlable", the SDK should translate it to some atoms that can be matched. If returning a description, IMO it should be raised

Comment (LVala) [reply lib/jellyfish/utils.ex:8]: So for the time being, I left it as it was.

File: README.md [hunk 12]
```diff
@@ -1,27 +1,51 @@
...
 ## Usage
 
-TODO
+Make API calls to Jellyfish:
+
+```elixir
+client = Jellyfish.Client.new("http://address-of-your-server.com")
+
+# Create room
+{:ok, room} = Jellyfish.Room.create(client, max_peers: 10)
```
Comment (mickel8) [inline README.md:33]: IMO, creating a room should be on the client level. In other case you create a client and you don't know what's next.

I belive that the API should guide user on what is possible so I would think about it as:

1. client is the main module so let's create a new client
2. ok what can I do on client? I can create a room or, in the future, open WS for monitoring so let's create a room
3. now I have a room, what can I do on room? I can add or delete a peer/component

I am also not sure about returning room state after creating it. We don't return a new room state when we add a new peer so you still have to fetch room state manually.

Comment (LVala) [reply README.md:33]: De facto, `Client` represents single instance of Jellyfish, so maybe we should move the functions from `Client` to `Jellyfish` scope, then you would call `Jellyfish.new/1` or `Jellyfish.register/1` to register Jellyfish instance, `Jellyfish.create_room` etc., then I think it would look coherent. I thought of `Client` as of like standalone thing, as if it was something used to communicate with Jellyfish, moving these functions there doesn't seem right somehow.

**Review by mickel8 — APPROVED** [review]

File: lib/jellyfish/sdk.ex [hunk 13]
```diff
@@ -0,0 +1,33 @@
+defmodule Jellyfish.SDK.Client do
+  @moduledoc """
+  This module defines a `t:Jellyfish.SDK.Client.t/0` that stores all of the data related to SDK state.
+  """
+  @typedoc """
+  * `http_request` - stores information used to make HTTP requests
+  """
+  @type t :: %__MODULE__{
```
[the remark this answers is not in the document]
Comment (daniel-jodlos) [reply lib/jellyfish/sdk.ex]: Bump. No answer, type still not opaque.

**Review by daniel-jodlos — APPROVED** [review]

**Review by Rados13 — APPROVED** [review]

File: README.md [hunk 14]
```diff
@@ -1,27 +1,51 @@
...
+
+room.id
+# => "8878cd13-99a6-40d6-8d7e-8da23d803dab"
+
+# Add peer
+{:ok, peer} = Jellyfish.Room.add_peer(client, room.id, "webrtc") # pass room id and type of peer
+
+# Delete peer
+:ok = Jellyfish.Room.delete_peer(client, room.id, peer.id)
+```
```
Comment (bblaszkow06) [inline README.md:43]: Continuing the discussion from https://github.com/jellyfish-dev/server_sdk_elixir/pull/1#discussion_r1130218565 I think I need to stir things up a bit :wink:

I disagree that functions such as `create_room` should be on a client level. In the SDKs I've checked there are mainly 2 approaches for very simple APIs, the functions making requests are in top-level module e.g. https://hexdocs.pm/configcat/ConfigCat.html#get_value/4. Modules like `Client` or `Connection` are used for setting up connection parameters, sometimes providing a basic, low-level API used by other modules.

Often the client is then passed (sometimes implicitly by a global config) to functions specific for "object" manipulated by the API. Here are some examples:

 ```elixir
 # aws-elixir
client = AWS.Client.create("your-access-key-id", "your-secret-access-key", "us-east-1")
{:ok, result, resp} = AWS.Kinesis.list_streams(client, %{})
AWS.S3.get_object(client, "bucket-name-here", "object-key-here")

# amqp
{:ok, conn} = AMQP.Connection.open()
{:ok, chan} = AMQP.Channel.open(conn)
AMQP.Queue.declare(chan, "test_queue")
AMQP.Exchange.declare(chan, "test_exchange")

# stripity-stripe
# connection is setup globally and includes HTTP connection pool
{:ok, setup_intent} = Stripe.SetupIntent.create(%{})
{:ok, stripe_customer} = Stripe.Customer.create(new_customer)
{:ok, _result} = Stripe.PaymentMethod.attach(%{customer: stripe_customer.id, payment_method: payment_method_id})
{:ok, charge} = Stripe.PaymentIntent.create(%{
  amount: cents_int,
  currency: "USD",
  customer: stripe_customer.id,
  payment_method: payment_method_id,
  off_session: true,
  confirm: true
})

# extwilio
# It's a bit of a mess, but quoting docs:
# If you're a user, take a look at the various Resource modules, such as
# [ExTwilio.Account](https://hexdocs.pm/ex_twilio/ExTwilio.Account.html),
# [ExTwilio.Call](https://hexdocs.pm/ex_twilio/ExTwilio.Call.html).
#
# If you want to learn more about how ExTwilio works internally, take a gander at:

# [ExTwilio.Api](https://hexdocs.pm/ex_twilio/ExTwilio.Api.html)
# [ExTwilio.Resource](https://hexdocs.pm/ex_twilio/ExTwilio.Resource.html)
# [ExTwilio.Parser](https://hexdocs.pm/ex_twilio/ExTwilio.Parser.html)

# ExForce - salesforce API
{:ok, %{instance_url: instance_url} = oauth_response} =
  ExForce.OAuth.get_token(
    "https://login.salesforce.com",
    grant_type: "password",
    client_id: "client_id",
    client_secret: "client_secret",
    username: "username",
    password: "password" <> "security_token"
  )

{:ok, version_maps} = ExForce.versions(instance_url)
latest_version = version_maps |> Enum.map(&Map.fetch!(&1, "version")) |> List.last()

client = ExForce.build_client(oauth_response, api_version: latest_version)

names =
  ExForce.query_stream(client, "SELECT Name FROM Account")
  |> Stream.map(&Map.fetch!(&1.data, "Name"))
  |> Stream.take(50)
  |> Enum.to_list()
```

So, considering that we have `Jellyfish.Room` the creation should be there. That's also consistent with a REST-style resource.

EDIT: I seem to agree with @daniel-jodlos on this (https://github.com/jellyfish-dev/server_sdk_elixir/pull/1#discussion_r1116584342)

File: lib/jellyfish/client.ex [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    * `address` - url or IP address of the Jellyfish server instance
+  """
+  @spec new(String.t()) :: t()
+  def new(address) do
+    middleware = [
+      {Tesla.Middleware.BaseUrl, address},
+      Tesla.Middleware.JSON
+    ]
+
+    adapter = Tesla.Adapter.Hackney
```
Comment (bblaszkow06) [inline lib/jellyfish/client.ex:51]: 1. This could be configurable
2. IMO the standard now is mint or finch, but I guess it's not as important now and can be handled separately

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]
TBC

File: lib/jellyfish/client.ex [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+  alias Jellyfish.Exception.ResponseStructureError
+  alias Jellyfish.{Room, Utils}
+  alias Tesla.Env
+
+  @enforce_keys [
+    :http_client
+  ]
+  defstruct @enforce_keys
+
+  @type t :: %__MODULE__{
```
[the remark this answers is not in the document]
Comment (LVala) [reply lib/jellyfish/client.ex:33]: My bad, I didn't know that `@opaque` was a thing in Elixir, I thought you just used fancy words for it not having typedoc, but unfortunately it makes it opaque also for other modules inside the sdk, where we use it as a parameter and dialyzer is not happy with that. Similar client modules in `tesla` or `elixir-aws` are specified as `@type`, so left it as it is here.

File: README.md (hunk 14 above)
Comment (LVala) [reply README.md:43]: Ok, I reverted it to how it was before moving the logic to `Client`, so now the API looks like this:
```elixir
Jellyfish.Client.new
Jellyfish.Room.create
Jellyfish.Room.delete
Jellyfish.Room.list
Jellyfish.Room.get_by_id
Jellyfish.Room.add_peer
Jellyfish.Room.remove_peer
Jellyfish.Room.add_component
Jellyfish.Room.remove_component
```
But probably it would be more REST-like if we did
```elixir
Jellyfish.Room.list -> Jellyfish.list_rooms
Jellyfish.Room.{Peer, Component}.delete
```
and maybe (?)
```elixir
Jellyfish.Room.get_by_id -> Jellyfish.Room.get
```

And moved `Jellyfish.{Peer, Component}` to `Jellyfish.Room.{Peer, Component}`.

What do you think?

File: lib/jellyfish/client.ex (hunk 15 above)
Comment (LVala) [reply lib/jellyfish/client.ex]: I added a way to configure it via `config, :jellyfish, tesla_adapter: _`, I considered doing via parameter to `Jellyfish.Client.new`, but is seemed unnecessary (probably not many users will be interested in changing the http backend and even less (if any) will want to use different backends for different instances of Jellyfish)

File: README.md (hunk 14 above)
Comment (bblaszkow06) [reply README.md]: > `Jellyfish.Room.get_by_id -> Jellyfish.Room.get`

This is the most obvious change for me, getting by id is the default here

>`Jellyfish.Room.list -> Jellyfish.list_rooms`

I don't think it's necessary to move it to the top level. If anything, I'd consider making that `list_all` or `get_all`, but just `list` is fine as well

>`Jellyfish.Room.delete_(peer | component) -> Jellyfish.Room.{Peer, Component}.delete`
>And moved `Jellyfish.{Peer, Component}` to `Jellyfish.Room.{Peer, Component}`.

This puzzles me the most. On one hand, peers and components don't exist outside of the room. On the other hand, they are unique enough (there are no other types of peer and component that would be under different modules) and deeply nested modules aren't convenient to use. In Membrane, we went with more nesting and after a while, we were refactoring to flatten the modules.

Summing up, I'd stick to `Jellyfish.Room.delete_X` and `Jellyfish.Peer` but I'm not entirely certain about that.

File: lib/jellyfish/client.ex [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+  config :jellyfish, tesla_adapter: Tesla.Adapter.Hackney
+
+  # mix.exs
+  defp deps do
+    [
+      {:hackney, "~> 1.10"}
+    ]
+  end
+  ```
+  For the list of supported Tesla adapters refer to [Tesla docs](https://hexdocs.pm/tesla/readme.html#adapters).
```
Comment (bblaszkow06) [inline lib/jellyfish/client.ex:17]: This differs from the actual default value in code.

File: lib/jellyfish/component.ex [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    :type
+  ]
+  defstruct @enforce_keys
+
+  @typedoc """
+  Id of the component, unique within the room.
+  """
+  @type id :: String.t()
+
+  # TODO change links do docs to proper ones (here and in moduledoc)
```
Comment (bblaszkow06) [inline lib/jellyfish/component.ex:20]: The docs are deployed now: https://jellyfish-dev.github.io/jellyfish-docs/

File: lib/jellyfish/component.ex [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+  @typedoc """
+  Type of the component.
+
+  For more information refer to [Jellyfish documentation](https://www.membrane.stream).
+  """
+  @type type :: String.t()
+
+  @typedoc """
+  Type describing component options.
+  For the list of available options, please refer to the component's documentation.
```
Comment (bblaszkow06) [inline lib/jellyfish/component.ex:30]: I'm not sure what documentation I should refer to. I think the docs here should include that info, especially since there might be some mapping required (e.g. option named `someOption` in JSON might become an atom `some_option` here

File: lib/jellyfish/room.ex [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+      } ->
+        %Component{
+          id: id,
+          type: type
+        }
+
+      _other ->
+        raise ResponseStructureError
+    end
+  end
```
Comment (bblaszkow06) [inline lib/jellyfish/room.ex:246]: IMO these functions could be moved to appropriate struct modules, as e.g. `Jellyfish.Peer.from_json` or `Jellyfish.Peer.parse_response`

File: mix.exs [hunk 21]
```diff
@@ -37,10 +48,15 @@ defmodule Membrane.Template.Mixfile do
 
   defp deps do
     [
-      {:membrane_core, "~> 0.10.0"},
+      {:tesla, "~> 1.5"},
+      {:mint, "~> 1.0"},
```
Comment (bblaszkow06) [inline mix.exs:52]: If this is swappable, it should be probably optional and you'd have to recommend putting the appropriate HTTP client in deps by the user, but we can skip that for now. Just make sure there's a task for it :wink:

File: lib/jellyfish/client.ex (hunk 17 above)
Comment (LVala) [reply lib/jellyfish/client.ex:17]: Well, it was supposed to be an example of how the default adapter can be changed, as it's stated in the line above: `By default, Mint adapter for Tesla is used to make HTTP requests, but it can be changed:`

File: lib/jellyfish/component.ex (hunk 19 above)
Comment (LVala) [reply lib/jellyfish/component.ex:30]: I'll link it to docs. @daniel-jodlos suggested [here](https://github.com/jellyfish-dev/server_sdk_elixir/pull/1#discussion_r1122607967), against that, but I think we can change the `@type options :: Keyword.t()` to be more specific and include the actual option atoms. Then the mapping between atoms and JSON values wouldn't be a problem.

Comment (daniel-jodlos) [reply lib/jellyfish/component.ex]: Yeah, it depends how extandable Jellyfish would be. I heard some plans to allow adding custom components, and we would need to handle its options in this spec as well. But if we don't plan on adding any, we can list all options here. Or we can can do `[option()] | Keyword.t()` to both provide documentation and cover user-defined endpoints.

**Review by bblaszkow06 — APPROVED** [review]
